File: src/Api/Services/PostService.cs
using Microsoft.EntityFrameworkCore;
using PostGenerator.Api.Data;
using PostGenerator.Api.Models;
using PostGenerator.Core;

namespace PostGenerator.Api.Services;

public class PostService : IPostService
{
    private const int UpcomingTake = 10;

    private readonly AppDbContext _db;
    private readonly IImageService _imageService;

    public PostService(AppDbContext db, IImageService imageService)
    {
        _db = db;
        _imageService = imageService;
    }

    public async Task<(IReadOnlyList<PostDto> Items, int TotalCount)> ListAsync(int userId, IReadOnlyList<string>? platforms, IReadOnlyList<string>? statuses, DateTime? from, DateTime? to, int? skip = null, int? take = null, CancellationToken cancellationToken = default)
    {
        var q = _db.Posts.AsNoTracking().Where(p => p.UserId == userId);
        if (platforms is { Count: > 0 })
        {
            var platformSet = platforms
                .Select(s => Enum.TryParse<PostPlatform>(s, true, out var pl) ? pl : (PostPlatform?)null)
                .Where(x => x.HasValue)
                .Select(x => x!.Value)
                .ToHashSet();
            if (platformSet.Count > 0)
                q = q.Where(p => platformSet.Contains(p.Platform));
        }
        if (statuses is { Count: > 0 })
        {
            var statusSet = statuses
                .Select(s => Enum.TryParse<PostStatus>(s, true, out var st) ? st : (PostStatus?)null)
                .Where(x => x.HasValue)
                .Select(x => x!.Value)
                .ToHashSet();
            if (statusSet.Count > 0)
                q = q.Where(p => statusSet.Contains(p.Status));
        }
        if (from.HasValue) q = q.Where(p => p.ScheduledAt >= from || (p.ScheduledAt == null && p.CreatedAt >= from));
        if (to.HasValue) q = q.Where(p => p.ScheduledAt <= to);

        var ordered = q.OrderBy(p => p.ScheduledAt ?? DateTime.MaxValue).ThenBy(p => p.CreatedAt);
        var projected = ordered.Select(p => new PostDto(p.Id, p.UserId, p.TopicSummary, p.Platform.ToString(), p.Status.ToString(), p.ScheduledAt, p.PublishedAt, p.ExternalPostId, p.ViewsCount, p.LikesCount, p.CommentsCount, p.LastEngagementFetchedAt, p.Content, p.Script, p.ImageUrl, p.MetadataJson, p.Tone, p.Length, p.CreatedAt, p.UpdatedAt));

        int totalCount;
        List<PostDto> items;
        if (skip.HasValue || take.HasValue)
        {
            totalCount = await ordered.CountAsync(cancellationToken);
            var paged = projected;
            if (skip.HasValue) paged = paged.Skip(skip.Value);
            if (take.HasValue) paged = paged.Take(take.Value);
            items = await paged.ToListAsync(cancellationToken);
        }
        else
        {
            items = await projected.ToListAsync(cancellationToken);
            totalCount = items.Count;
        }
        return (items, totalCount);
    }

    public async Task<PostDto?> GetByIdAsync(int userId, int postId, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.AsNoTracking().FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        return post == null ? null : PostMappings.ToDto(post);
    }

    public async Task<PostDto> CreateAsync(int userId, CreatePostRequest request, CancellationToken cancellationToken = default)
    {
        if (!Enum.TryParse<PostPlatform>(request.Platform, true, out var platform))
            throw new ArgumentException("Invalid platform", nameof(request));

        var post = new Post
        {
            UserId = userId,
            TopicSummary = request.TopicSummary ?? "",
            Platform = platform,
            Status = PostStatus.Draft,
            ScheduledAt = request.ScheduledAt,
            Content = request.Content ?? "",
            Script = request.Script,
            ImageUrl = request.ImageUrl,
            MetadataJson = request.MetadataJson,
            Tone = request.Tone,
            Length = request.Length,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };
        _db.Posts.Add(post);
        await _db.SaveChangesAsync(cancellationToken);
        return PostMappings.ToDto(post);
    }

    public async Task<PostDto?> UpdateAsync(int userId, int postId, UpdatePostRequest request, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        if (post == null) return null;

        if (request.TopicSummary != null) post.TopicSummary = request.TopicSummary;
        if (request.Content != null) post.Content = request.Content;
        if (request.Script != null) post.Script = request.Script;
        if (request.ImageUrl != null) post.ImageUrl = request.ImageUrl;
        if (request.MetadataJson != null) post.MetadataJson = request.MetadataJson;
        if (request.Tone != null) post.Tone = request.Tone;
        if (request.Length != null) post.Length = request.Length;
        if (request.ScheduledAt.HasValue) post.ScheduledAt = request.ScheduledAt;
        if (request.Status != null && Enum.TryParse<PostStatus>(request.Status, true, out var st)) post.Status = st;
        post.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        return PostMappings.ToDto(post);
    }

    public async Task<bool> DeleteAsync(int userId, int postId, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        if (post == null) return false;
        _db.Posts.Remove(post);
        await _db.SaveChangesAsync(cancellationToken);
        return true;
    }

    // public async Task<PostDto?> GenerateImageAsync(int userId, int postId, string? prompt, CancellationToken cancellationToken = default)
    // {
    //     var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
    //     if (post == null) return null;

    //     var effectivePrompt = !string.IsNullOrWhiteSpace(prompt)
    //         ? prompt
    //         : $"Social media image for post: {post.Content[..Math.Min(500, post.Content.Length)]}";
    //     var imageUrl = await _imageService.GenerateAndUploadAsync(effectivePrompt, $"post-{post.Id}-{Guid.NewGuid():N}.png", cancellationToken);
    //     if (imageUrl == null) return null;

    //     post.ImageUrl = imageUrl;
    //     post.UpdatedAt = DateTime.UtcNow;
    //     await _db.SaveChangesAsync(cancellationToken);
    //     return PostMappings.ToDto(post);
    // }

    public async Task<PostDto?> SetPostImageUrlAsync(int userId, int postId, string? imageUrl, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        if (post == null) return null;

        post.ImageUrl = imageUrl;
        post.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        return PostMappings.ToDto(post);
    }

    public async Task<PostDto?> PublishNowAsync(int userId, int postId, CancellationToken cancellationToken = default)
    {
        var post = await _db.Posts.FirstOrDefaultAsync(p => p.Id == postId && p.UserId == userId, cancellationToken);
        if (post == null || post.Status != PostStatus.Draft) return null;

        post.Status = PostStatus.Published;
        post.PublishedAt = DateTime.UtcNow;
        post.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        return PostMappings.ToDto(post);
    }

    public async Task<DashboardStatsDto> GetDashboardStatsAsync(int userId, CancellationToken cancellationToken = default)
    {
        var now = DateTime.UtcNow;
        var baseQ = _db.Posts.AsNoTracking().Where(p => p.UserId == userId);

        var total = await baseQ.CountAsync(cancellationToken);
        var draftCount = await baseQ.CountAsync(p => p.Status == PostStatus.Draft, cancellationToken);
        var scheduledCount = await baseQ.CountAsync(p => p.Status == PostStatus.Scheduled, cancellationToken);
        var publishedCount = await baseQ.CountAsync(p => p.Status == PostStatus.Published, cancellationToken);
        var failedCount = await baseQ.CountAsync(p => p.Status == PostStatus.Failed, cancellationToken);

        var byPlatformRows = await baseQ
            .GroupBy(p => p.Platform)
            .Select(g => new { Platform = g.Key, Count = g.Count() })
            .OrderByDescending(x => x.Count)
            .ToListAsync(cancellationToken);
        var byPlatform = byPlatformRows.Select(x => new PostsByPlatformDto(x.Platform.ToString(), x.Count)).ToList();

        var upcomingPosts = await baseQ
            .Where(p => p.Status == PostStatus.Scheduled && p.ScheduledAt != null && p.ScheduledAt >= now)
            .OrderBy(p => p.ScheduledAt)
            .Take(UpcomingTake)
            .Select(p => new UpcomingPostDto(p.Id, p.Platform.ToString(), p.ScheduledAt!.Value, p.TopicSummary))
            .ToListAsync(cancellationToken);

        var mostRecentPublished = await baseQ
            .Where(p => p.Status == PostStatus.Published && p.PublishedAt != null)
            .OrderByDescending(p => p.PublishedAt)
            .Select(p => new PostDto(p.Id, p.UserId, p.TopicSummary, p.Platform.ToString(), p.Status.ToString(), p.ScheduledAt, p.PublishedAt, p.ExternalPostId, p.ViewsCount, p.LikesCount, p.CommentsCount, p.LastEngagementFetchedAt, p.Content, p.Script, p.ImageUrl, p.MetadataJson, p.Tone, p.Length, p.CreatedAt, p.UpdatedAt))
            .FirstOrDefaultAsync(cancellationToken);

        return new DashboardStatsDto(
            total,
            draftCount,
            scheduledCount,
            publishedCount,
            failedCount,
            byPlatform,
            upcomingPosts,
            mostRecentPublished);
    }
}

File: src/lib/services/post-service.ts
import { db } from "@/db/drizzle";
import {
  posts,
  type Post,
  type PostPlatform,
  type PostStatus,
} from "@/db/schema";
import { eq, and, inArray, asc } from "drizzle-orm";
import type { NewPost } from "@/db/schema";

const PLATFORMS: PostPlatform[] = [
  "LinkedIn",
  "Skool",
  "Instagram",
  "Bluesky",
  "Facebook",
  "TikTok",
];
const STATUSES: PostStatus[] = ["Draft", "Scheduled", "Published", "Failed"];

function isValidPlatform(s: string): s is PostPlatform {
  return PLATFORMS.includes(s as PostPlatform);
}
function isValidStatus(s: string): s is PostStatus {
  return STATUSES.includes(s as PostStatus);
}

export interface PostDto {
  id: number;
  userId: number;
  topicSummary: string;
  platform: string;
  status: string;
  scheduledAt: string | null;
  publishedAt: string | null;
  externalPostId: string | null;
  viewsCount: number | null;
  likesCount: number | null;
  commentsCount: number | null;
  lastEngagementFetchedAt: string | null;
  content: string;
  script: string | null;
  imageUrl: string | null;
  metadataJson: string | null;
  tone: string | null;
  length: string | null;
  createdAt: string;
  updatedAt: string;
}

function toDto(p: Post): PostDto {
  return {
    id: p.id,
    userId: p.userId,
    topicSummary: p.topicSummary ?? "",
    platform: p.platform,
    status: p.status,
    scheduledAt: p.scheduledAt?.toISOString() ?? null,
    publishedAt: p.publishedAt?.toISOString() ?? null,
    externalPostId: p.externalPostId,
    viewsCount: p.viewsCount,
    likesCount: p.likesCount,
    commentsCount: p.commentsCount,
    lastEngagementFetchedAt: p.lastEngagementFetchedAt?.toISOString() ?? null,
    content: p.content,
    script: p.script,
    imageUrl: p.imageUrl,
    metadataJson: p.metadataJson,
    tone: p.tone,
    length: p.length,
    createdAt: p.createdAt.toISOString(),
    updatedAt: p.updatedAt.toISOString(),
  };
}

export interface ListResult {
  items: PostDto[];
  totalCount: number;
}

export async function list(
  userId: number,
  opts: {
    platforms?: string[];
    statuses?: string[];
    from?: Date;
    to?: Date;
    skip?: number;
    take?: number;
  }
): Promise<ListResult> {
  const conditions = [eq(posts.userId, userId)];

  const platformSet = opts.platforms?.filter(isValidPlatform) ?? [];
  if (platformSet.length > 0) {
    conditions.push(inArray(posts.platform, platformSet));
  }

  const statusSet = opts.statuses?.filter(isValidStatus) ?? [];
  if (statusSet.length > 0) {
    conditions.push(inArray(posts.status, statusSet));
  }

  const whereClause = conditions.length === 1 ? conditions[0] : and(...conditions);

  let rows = await db
    .select()
    .from(posts)
    .where(whereClause)
    .orderBy(asc(posts.scheduledAt), asc(posts.createdAt));

  if (opts.from) {
    rows = rows.filter(
      (r) =>
        (r.scheduledAt && r.scheduledAt >= opts.from!) ||
        (!r.scheduledAt && r.createdAt >= opts.from!)
    );
  }
  if (opts.to) {
    rows = rows.filter((r) => !r.scheduledAt || r.scheduledAt <= opts.to!);
  }

  const totalCount = rows.length;
  const skip = opts.skip ?? 0;
  const take = opts.take ?? 20;
  const paged = rows.slice(skip, skip + take);

  return { items: paged.map(toDto), totalCount };
}

export async function getById(
  userId: number,
  postId: number
): Promise<PostDto | null> {
  const [p] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);
  return p ? toDto(p) : null;
}

export interface CreatePostRequest {
  topicSummary: string;
  platform: string;
  content?: string | null;
  script?: string | null;
  imageUrl?: string | null;
  metadataJson?: string | null;
  tone?: string | null;
  length?: string | null;
  scheduledAt?: Date | null;
}

export async function create(
  userId: number,
  req: CreatePostRequest
): Promise<PostDto> {
  if (!isValidPlatform(req.platform)) {
    throw new Error("Invalid platform");
  }

  const [inserted] = await db
    .insert(posts)
    .values({
      userId,
      topicSummary: req.topicSummary ?? "",
      platform: req.platform,
      status: "Draft",
      content: req.content ?? "",
      script: req.script,
      imageUrl: req.imageUrl,
      metadataJson: req.metadataJson,
      tone: req.tone,
      length: req.length,
      scheduledAt: req.scheduledAt ? new Date(req.scheduledAt) : null,
    } as NewPost)
    .returning();

  return toDto(inserted!);
}

export interface UpdatePostRequest {
  topicSummary?: string | null;
  content?: string | null;
  script?: string | null;
  imageUrl?: string | null;
  metadataJson?: string | null;
  tone?: string | null;
  length?: string | null;
  scheduledAt?: Date | null;
  status?: string | null;
}

export async function update(
  userId: number,
  postId: number,
  req: UpdatePostRequest
): Promise<PostDto | null> {
  const [existing] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);
  if (!existing) return null;

  const updates: Partial<Post> = { updatedAt: new Date() };
  if (req.topicSummary !== undefined) updates.topicSummary = req.topicSummary;
  if (req.content !== undefined && req.content != null) updates.content = req.content;
  if (req.script !== undefined) updates.script = req.script;
  if (req.imageUrl !== undefined) updates.imageUrl = req.imageUrl;
  if (req.metadataJson !== undefined) updates.metadataJson = req.metadataJson;
  if (req.tone !== undefined) updates.tone = req.tone;
  if (req.length !== undefined) updates.length = req.length;
  if (req.scheduledAt !== undefined)
    updates.scheduledAt = req.scheduledAt ? new Date(req.scheduledAt) : null;
  if (
    req.status !== undefined &&
    req.status !== null &&
    isValidStatus(req.status)
  )
    updates.status = req.status;

  const [updated] = await db
    .update(posts)
    .set(updates as Partial<Post>)
    .where(eq(posts.id, postId))
    .returning();

  return updated ? toDto(updated) : null;
}

export async function remove(
  userId: number,
  postId: number
): Promise<boolean> {
  const [existing] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);
  if (!existing) return false;

  await db.delete(posts).where(eq(posts.id, postId));
  return true;
}

export async function setPostImageUrl(
  userId: number,
  postId: number,
  imageUrl: string | null
): Promise<PostDto | null> {
  const [existing] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);
  if (!existing) return null;

  const [updated] = await db
    .update(posts)
    .set({ imageUrl, updatedAt: new Date() })
    .where(eq(posts.id, postId))
    .returning();

  return updated ? toDto(updated) : null;
}

export async function publishNow(
  userId: number,
  postId: number
): Promise<PostDto | null> {
  const [existing] = await db
    .select()
    .from(posts)
    .where(and(eq(posts.id, postId), eq(posts.userId, userId)))
    .limit(1);
  if (!existing || existing.status !== "Draft") return null;

  const now = new Date();
  const [updated] = await db
    .update(posts)
    .set({ status: "Published", publishedAt: now, updatedAt: now })
    .where(eq(posts.id, postId))
    .returning();

  return updated ? toDto(updated) : null;
}

export interface DashboardStats {
  totalPosts: number;
  draftCount: number;
  scheduledCount: number;
  publishedCount: number;
  failedCount: number;
  byPlatform: { platform: string; count: number }[];
  upcomingPosts: {
    id: number;
    platform: string;
    scheduledAt: string;
    topicSummary: string;
  }[];
  mostRecentPublished: PostDto | null;
}

const UPCOMING_TAKE = 10;

export async function getDashboardStats(userId: number): Promise<DashboardStats> {
  const all = await db
    .select()
    .from(posts)
    .where(eq(posts.userId, userId));

  const totalPosts = all.length;
  const draftCount = all.filter((p) => p.status === "Draft").length;
  const scheduledCount = all.filter((p) => p.status === "Scheduled").length;
  const publishedCount = all.filter((p) => p.status === "Published").length;
  const failedCount = all.filter((p) => p.status === "Failed").length;

  const platformCounts = new Map<string, number>();
  for (const p of all) {
    platformCounts.set(p.platform, (platformCounts.get(p.platform) ?? 0) + 1);
  }
  const byPlatform = Array.from(platformCounts.entries())
    .map(([platform, count]) => ({ platform, count }))
    .sort((a, b) => b.count - a.count);

  const now = new Date();
  const upcoming = all
    .filter(
      (p) =>
        p.status === "Scheduled" &&
        p.scheduledAt &&
        p.scheduledAt >= now
    )
    .sort((a, b) => (a.scheduledAt!.getTime() - b.scheduledAt!.getTime()))
    .slice(0, UPCOMING_TAKE)
    .map((p) => ({
      id: p.id,
      platform: p.platform,
      scheduledAt: p.scheduledAt!.toISOString(),
      topicSummary: p.topicSummary ?? "",
    }));

  const mostRecentPublished = all
    .filter((p) => p.status === "Published" && p.publishedAt)
    .sort((a, b) => b.publishedAt!.getTime() - a.publishedAt!.getTime())[0];

  return {
    totalPosts,
    draftCount,
    scheduledCount,
    publishedCount,
    failedCount,
    byPlatform,
    upcomingPosts: upcoming,
    mostRecentPublished: mostRecentPublished
      ? toDto(mostRecentPublished)
      : null,
  };
}
